### PR TITLE
Support `::crate` in paths

### PR DESCRIPTION
**Pull request by petrochenkov**
cc https://github.com/rust-lang/rust/issues/45477
Fixes https://github.com/rust-lang/rust/issues/45229

### CODE REVIEW

File: src/libsyntax/ast.rs [hunk 1]
```diff
@@ -97,9 +97,12 @@ impl Path {
     }
 
     pub fn default_to_global(mut self) -> Path {
```
Comment (nikomatsakis) [inline src/libsyntax/ast.rs]: Pre-existing, but I think a comment on this function would be great, just showing the transformations it does (and maybe explaining the reason for its existence).

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -3992,7 +3992,9 @@ impl<'a> Parser<'a> {
         // Starts like a simple path, but not a union item.
         } else if self.token.is_path_start() &&
                   !self.token.is_qpath_start() &&
-                  !self.is_union_item() {
+                  !self.is_union_item() &&
+                  !(self.token.is_keyword(keywords::Crate) &&
+                    self.look_ahead(1, |t| t != &token::ModSep)) {
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: Nit: I feel like some comments would be nice here too. In fact, I personally am a bit confused about what's going on here. =) Lots of nested nots! 

```
// Parse any path `foo::bar`. But be careful of various corner cases involving contextual keywords:
//
// - `union Foo { .. }`, a union item. This gets parsed below. Note that `union::foo` is a path.
// - Note that `crate::foo` *is* a path, but `crate` followed by any other token is not. (??)
```

I'm not entirely clear I guess on what is happening. =)

File: src/test/compile-fail/crate-path-non-absolute.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
```
Comment (nikomatsakis) [inline src/test/compile-fail/crate-path-non-absolute.rs]: Can we please collect the tests for this feature in some kind of directory named after the rfc? I would suggest `{compile-fail,run-pass,ui}/rfc2126-crate-paths`.

File: src/test/compile-fail/crate-visibility-ambiguity.rs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// except according to those terms.
+
+#![feature(crate_in_paths)]
+#![feature(crate_visibility_modifier)]
+
+mod m {
+    pub struct Z;
+    pub struct S1(crate (::m::Z)); // OK
+    pub struct S2(::crate ::m::Z); // OK
+    pub struct S3(crate ::m::Z); //~ ERROR undeclared type or module `crate`
```
Comment (nikomatsakis) [inline src/test/compile-fail/crate-visibility-ambiguity.rs]: Huh, this error seems to indicate that it is parsing a path `crate::m::Z` which then fails to resolve... is that correct? If so, that seems sort of wrong, right? [Today at least you get an error](https://play.rust-lang.org/?gist=1015758e53b485dba64b567b06cc29de&version=stable), and if we were going to favor *any* interpretation here, I would expect us to treat `crate` as a visibility modifier (wasn't that the whole point of accepting `::crate` only?).

File: src/test/compile-fail/use-super-global-path.rs [hunk 5]
```diff
@@ -14,10 +14,10 @@ struct S;
 struct Z;
 
 mod foo {
-    use ::super::{S, Z}; //~ ERROR global paths cannot start with `super`
+    use ::super::{S, Z}; //~ ERROR unresolved import `super`
 
     pub fn g() {
-        use ::super::main; //~ ERROR global paths cannot start with `super`
+        use ::super::main; //~ ERROR unresolved import `super`
```
Comment (nikomatsakis) [inline src/test/compile-fail/use-super-global-path.rs]: Why the change to the error message here? Seems less good now.

File: src/test/run-pass/crate-path-absolute.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(crate_in_paths)]
+
+use crate::m::f;
```
Comment (nikomatsakis) [inline src/test/run-pass/crate-path-absolute.rs]: Can we also test this from another module besides the root module?

File: src/test/run-pass/crate-path-absolute.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+mod m {
+    pub fn f() {}
+    pub fn g() {}
+
+    // OK, visibilities are implicitly absolute like imports
+    pub(in crate::m) struct S;
+}
+
+fn main() {
+    f();
```
Comment (nikomatsakis) [inline src/test/run-pass/crate-path-absolute.rs]: Not that I think it's *really* needed, bu this test doesn't demonstrate that things resolved correctly in any way. Perhaps have `f` and `g` return values, and we can assert that you get the right thing back?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
I left various requests around the tests, but my biggest concern is that `crate::foo::bar` seems to parse in contexts where I would not expect it to.

**Comment by shepmaster** [comment]
Hi from triage @petrochenkov — looks like you've got some grade-A feedback; will you be able to act on it soon?

File: src/test/compile-fail/crate-visibility-ambiguity.rs (hunk 4 above)
Comment (petrochenkov) [reply src/test/compile-fail/crate-visibility-ambiguity.rs]: Yes, `crate ::m::Z` is parsed as a single path, I explained this choice in https://github.com/rust-lang/rust/issues/45477#issuecomment-340595412.
FWIW, it's also the most conservative solution (since `crate::anything::else` is an error).

>wasn't that the whole point of accepting ::crate only?

I thought it was for teachability / symmetry with `::other_crate::x::y` (https://github.com/rust-lang/rust/issues/44660#issuecomment-330598910).

**Comment by petrochenkov** [comment]
>will you be able to act on it soon?

I'll address the comments tomorrow.

**Comment by petrochenkov** [comment]
Updated.

File: src/test/compile-fail/rfc-2126-crate-paths/crate-visibility-ambiguity.rs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// except according to those terms.
+
+#![feature(crate_in_paths)]
+#![feature(crate_visibility_modifier)]
+
+mod m {
+    pub struct Z;
+    pub struct S1(crate (::m::Z)); // OK
+    pub struct S2(::crate ::m::Z); // OK
+    pub struct S3(crate ::m::Z); //~ ERROR `crate` can only be used in absolute paths
```
Comment (nikomatsakis) [inline src/test/compile-fail/rfc-2126-crate-paths/crate-visibility-ambiguity.rs]: I'm still a bit surprised that this doesn't work out with `crate` being a visibility modifier. What is the reason for it not to parse that way?

File: src/test/compile-fail/use-super-global-path.rs [hunk 9]
```diff
@@ -14,10 +14,12 @@ struct S;
...
 
 mod foo {
-    use ::super::{S, Z}; //~ ERROR global paths cannot start with `super`
+    use ::super::{S, Z}; //~ ERROR unresolved import `super`
+                         //~^ NOTE global paths cannot start with `super`
 
     pub fn g() {
-        use ::super::main; //~ ERROR global paths cannot start with `super`
+        use ::super::main; //~ ERROR unresolved import `super`
+                           //~^ NOTE global paths cannot start with `super`
```
Comment (nikomatsakis) [inline src/test/compile-fail/use-super-global-path.rs]: The notes are definitely an improvement -- but "unresolved import `super`" still feels misleading to me. Among other things, `super` is a keyword and hence could not *possibly* be an import, right? Can we readily change from adding a note to altering the main message?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks nice. Two more questions. =)

File: src/test/compile-fail/use-super-global-path.rs (hunk 9 above)
Comment (petrochenkov) [reply src/test/compile-fail/use-super-global-path.rs]: Done.

File: src/test/compile-fail/rfc-2126-crate-paths/crate-visibility-ambiguity.rs (hunk 8 above)
Comment (petrochenkov) [reply src/test/compile-fail/rfc-2126-crate-paths/crate-visibility-ambiguity.rs]: I answered this earlier (https://github.com/rust-lang/rust/pull/45771#discussion_r150357994), but GitHub has hidden the message.

**Comment by nikomatsakis** [comment]
@petrochenkov 

Regarding the precedence of `crate`, I don't agree with your conclusion at the moment, but I'd rather not get stuck on it. Let's land the PR as is. I expect us to be revisiting various aspects of the syntax here.

**Comment by nikomatsakis** [comment]
r=me post rebase

**Comment by petrochenkov** [comment]
@nikomatsakis 
>I expect us to be revisiting various aspects of the syntax here in any case.

FWIW, I still hope this never ends up on stable.
